cli: schema validation of jinja template user-data (SC-1385)

blackboxsw opened the pull request:
## Proposed Commit Message
```
cli: schema validation of jinja template user-data

The CLI cloud-init schema now asserts that the leading header comment in user-data files is a valid user-data type. Raise an informative error otherwise about valid user-data types.

For user-data files declared with '## template: jinja', render those files firts sourcing jinja varibles from /run/cloud-init/instance-data.json or a new --instance-data parameter.

Once the jinja template is rendered, validate schema of the resulting #cloud-config user-data.

This branch also ensures any errors and deprecation warnings are unique.

LP: #1881925
```

## Additional Context
<!-- If relevant -->

## Test Steps
<!-- Please include any steps necessary to verify (and reproduce if
this is a bug fix) this change on a live deployed system,
including any necessary configuration files, user-data,
setup, and teardown. Scripts used may be attached directly to this PR. -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes
that apply. -->
 - [ ] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/contributing.html)
 - [ ] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any documentation accordingly

aciba90 reviewed (changes requested):
Overall, looks great, thanks!

I have fixed a couple of lint issues and simplified `SchemaValidationError.__init__` in the following patch:

```diff
diff --git a/cloudinit/config/schema.py b/cloudinit/config/schema.py
index 993feb148..b57e6c6c7 100644
--- a/cloudinit/config/schema.py
+++ b/cloudinit/config/schema.py
@@ -138,23 +138,25 @@ class SchemaValidationError(ValueError):
             ((flat.config.key, msg),)
         """
         message = ""
-        if schema_errors:
-            self.schema_errors = sorted(set(schema_errors))
-            message += _format_schema_problems(
-                self.schema_errors, prefix="Cloud config schema errors: "
-            )
-        else:
-            self.schema_errors = schema_errors
-        if schema_deprecations:
-            self.schema_deprecations = sorted(set(schema_deprecations))
+
+        def handle_problems(problems, prefix):
+            if not problems:
+                return problems
+            nonlocal message
             if message:
                 message += "\n\n"
-            message += _format_schema_problems(
-                self.schema_deprecations,
-                prefix="Cloud config schema deprecations: ",
-            )
-        else:
-            self.schema_deprecations = schema_deprecations
+            problems = sorted(list(set(problems)))
+            message += _format_schema_problems(problems, prefix=prefix)
+            return problems
+
+        self.schema_errors = handle_problems(
+            schema_errors,
+            prefix="Cloud config schema errors: ",
+        )
+        self.schema_deprecations = handle_problems(
+            schema_deprecations,
+            prefix="Cloud config schema deprecations: ",
+        )
         super().__init__(message)
 
     def has_errors(self) -> bool:
@@ -711,7 +713,7 @@ def validate_cloudconfig_file(
                         "But, content is not a jinja template",
                     )
                 ]
-            )
+            ) from None
         except JinjaLoadError as e:
             error(str(e), sys_exit=True)
         schema_position = "format-l2.c1"
@@ -754,7 +756,7 @@ def validate_cloudconfig_file(
         if annotate:
             print(
                 annotated_cloudconfig_file(
-                    {}, content, {}, schema_errors=errors
+                    {}, content, {}, schema_errors=schema_error.schema_errors
                 )
             )
         raise schema_error from e

```

blackboxsw commented:
> Overall, looks great, thanks!
> 
> I have fixed a couple of lint issues and simplified `SchemaValidationError.__init__` in the following patch:
> 
> ```diff
> diff --git a/cloudinit/config/schema.py b/cloudinit/config/schema.py
> index 993feb148..b57e6c6c7 100644
> --- a/cloudinit/config/schema.py
> +++ b/cloudinit/config/schema.py
> @@ -138,23 +138,25 @@ class SchemaValidationError(ValueError):
>              ((flat.config.key, msg),)
>          """
>          message = ""
> -        if schema_errors:
> -            self.schema_errors = sorted(set(schema_errors))
> -            message += _format_schema_problems(
> -                self.schema_errors, prefix="Cloud config schema errors: "
> -            )
> -        else:
> -            self.schema_errors = schema_errors
> -        if schema_deprecations:
> -            self.schema_deprecations = sorted(set(schema_deprecations))
> +
> +        def handle_problems(problems, prefix):
> +            if not problems:
> +                return problems
> +            nonlocal message
>              if message:
>                  message += "\n\n"
> -            message += _format_schema_problems(
> -                self.schema_deprecations,
> -                prefix="Cloud config schema deprecations: ",
> -            )
> -        else:
> -            self.schema_deprecations = schema_deprecations
> +            problems = sorted(list(set(problems)))
> +            message += _format_schema_problems(problems, prefix=prefix)
> +            return problems
> +
> +        self.schema_errors = handle_problems(
> +            schema_errors,
> +            prefix="Cloud config schema errors: ",
> +        )
> +        self.schema_deprecations = handle_problems(
> +            schema_deprecations,
> +            prefix="Cloud config schema deprecations: ",
> +        )
>          super().__init__(message)
>  
>      def has_errors(self) -> bool:
> @@ -711,7 +713,7 @@ def validate_cloudconfig_file(
>                          "But, content is not a jinja template",
>                      )
>                  ]
> -            )
> +            ) from None
>          except JinjaLoadError as e:
>              error(str(e), sys_exit=True)
>          schema_position = "format-l2.c1"
> @@ -754,7 +756,7 @@ def validate_cloudconfig_file(
>          if annotate:
>              print(
>                  annotated_cloudconfig_file(
> -                    {}, content, {}, schema_errors=errors
> +                    {}, content, {}, schema_errors=schema_error.schema_errors
>                  )
>              )
>          raise schema_error from e
> ```

Thanks @aciba90 for the review. Finally got back to updating this PR.

aciba90 reviewed (approved):
LGTM, thanks!
